### PR TITLE
[CONTINT-3435][PLD] Forward every query to the leader

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
This PR modifies the language detection handle to forward every request received by DCA followers to the leader.
It also changes the leader election `GetLeaderIP` function to cache the IP for a given pod name

### Motivation
Will be useful for cleanup and simplicity.

### Additional Notes
The cluster checks handler has a similar mechanism but it requires some work to extract the logic that handles the leader election from the logic that handles cluster checks.
For simplicity and to avoid unexpected breaking changes in the clc handlers, we chose a different approach.

### Possible Drawbacks / Trade-offs
`GetLeaderIP` will not return an error if the leader is not listed in the endpoints object. This function is only used in CLC runners.

### Describe how to test/QA your changes
1. Deploy multiple cluster agent with process language detection enabled and remote config enabled (for rbacs).

Helm:
```
datadog:
  remoteConfiguration:
    enabled: true
  env:
    - name: DD_TELEMETRY_ENABLED
      value: "true"
clusterAgent:
  env:
    - name: DD_LANGUAGE_DETECTION_ENABLED
      value: "true"
  replicas: 2
  admissionController:
    enabled: true
    mutateUnlabelled: true
    configMode: hostip
    remoteInstrumentation:
      enabled: true # important
```
3. Deploy a dummy python deployment
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: python-sleep-app
spec:
  replicas: 4
  selector:
    matchLabels:
      app: python-sleep-app
  template:
    metadata:
      labels:
        app: python-sleep-app
    spec:
      containers:
      - name: python-sleep
        image: python:3.9
        command: ["python"]
        args: ["-c", "import time; while True: time.sleep(5)"]
```
5. The telemetry can help to know how many requests were forwarded and how many were processed. 
```
❯ k -n datadog-agent-helm exec datadog-agent-linux-cluster-agent-6cc67c8d54-hrngs -- datadog-cluster-agent telemetry | grep -i forwarded_request
# HELP forwarded_requests Counter of requests forwarded to the dca leader
# TYPE forwarded_requests counter
forwarded_requests{handler="postDetectedLanguages"} 1
❯ k -n datadog-agent-helm exec datadog-agent-linux-cluster-agent-6cc67c8d54-hrngs -- agent status | grep -i follower
  Status: Follower, redirecting to leader at 10.244.1.27
  Status: Follower, cluster agent leader is: datadog-agent-linux-cluster-agent-6cc67c8d54-s9zkg
```

For the leader:
```
❯ k -n datadog-agent-helm exec datadog-agent-linux-cluster-agent-6cc67c8d54-s9zkg -- datadog-cluster-agent telemetry | grep -i forwarded_request
Defaulted container "cluster-agent" out of: cluster-agent, init-volume (init)
❯ k -n datadog-agent-helm exec datadog-agent-linux-cluster-agent-6cc67c8d54-s9zkg -- datadog-cluster-agent telemetry | grep -i dca
Defaulted container "cluster-agent" out of: cluster-agent, init-volume (init)
# HELP language_detection_dca_handler_ok_response Tracks the number the request was processed successfully
# TYPE language_detection_dca_handler_ok_response counter
language_detection_dca_handler_ok_response 1
# HELP language_detection_dca_handler_skipped_patch Tracks the number of times a patch was skipped because no new languages are detected
# TYPE language_detection_dca_handler_skipped_patch counter
language_detection_dca_handler_skipped_patch{namespace="default",owner_kind="Deployment",owner_name="python-sleep-app"} 1
```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: cmd/cluster-agent/subcommands/start/command.go [hunk 1]
```diff
@@ -168,8 +169,14 @@ func start(log log.Component, config config.Component, telemetry telemetry.Compo
 		}
 	}
 
+	// Setup the leader forwarder for language detection and cluster checks
+	var leaderForwarder *apidca.LeaderForwarder
```
Comment (AliDatadog) [inline cmd/cluster-agent/subcommands/start/command.go:173]: Here the leader engine still does not exist but I think we might not need it

File: cmd/cluster-agent/subcommands/start/command.go [hunk 2]
```diff
@@ -168,8 +169,14 @@ func start(log log.Component, config config.Component, telemetry telemetry.Compo
...
+	if pkgconfig.Datadog.GetBool("cluster_checks.enabled") || pkgconfig.Datadog.GetBool("language_detection.enabled") {
+		leaderForwarder = apidca.NewLeaderForwarder(pkgconfig.Datadog.GetInt("cluster_agent.cmd_port"), pkgconfig.Datadog.GetInt("cluster_agent.max_connections"))
```
Comment (AliDatadog) [inline cmd/cluster-agent/subcommands/start/command.go:175]: `cluster_agent.max_connections` will most likely be reached pretty fast and it might affect clc runners, should we have a separate leaderForwarder for both ? In that case we might have two connections between two DCA.

File: pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go [hunk 3]
```diff
@@ -243,13 +247,20 @@ func (le *LeaderEngine) GetLeader() string {
 
 // GetLeaderIP returns the IP the leader can be reached at, assuming its
 // identity is its pod name. Returns empty if we are the leader.
-// The result is not cached.
+// The result is cached and will not return an error if the leader does not exist anymore.
 func (le *LeaderEngine) GetLeaderIP() (string, error) {
```
Comment (AliDatadog) [inline pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go:251]: This is a breaking change with the current interface however I doubt that it will have a strong impact.
It used to always do an API Server Request to make sure that the leader is still up and running and retrieve its ip. Now it doesn't.
It means that:
- If the leader is down, it doesn't return an error when it should
- If the leader changed its IP (unlikely), it will return the old IP address.

File: cmd/cluster-agent/subcommands/start/command.go (hunk 1 above)
Comment (AliDatadog) [reply cmd/cluster-agent/subcommands/start/command.go:173]: This choice was made because both clc runners and pld handler need the leader forwarder. However, we don't want to duplicate connections.

File: cmd/cluster-agent/api/server.go [hunk 4]
```diff
@@ -46,7 +47,7 @@ var (
 )
 
 // StartServer creates the router and starts the HTTP server
-func StartServer(senderManager sender.DiagnoseSenderManager) error {
+func StartServer(senderManager sender.DiagnoseSenderManager, leaderForwarder *api.LeaderForwarder) error {
```
Comment (AliDatadog) [inline cmd/cluster-agent/api/server.go:50]: We pass down the same leader forwarder to both the CLC handler and the PLD handler. But we could have a global one or separate ones. Any opinion here ?

**Review by jhgilbert — APPROVED** [review]

File: pkg/util/kubernetes/apiserver/leaderelection/leaderelection_test.go [hunk 5]
```diff
@@ -409,40 +397,15 @@ func TestGetLeaderIPFollower_Lease(t *testing.T) {
...
-
-//nolint:revive // TODO(CAPP) Fix revive linter
-func (g *dummyGauge) Delete(tagsValue ...string) {}
-
-//nolint:revive // TODO(CAPP) Fix revive linter
-func (g *dummyGauge) WithValues(tagsValue ...string) telemetryComponent.SimpleGauge { return nil }
-
-//nolint:revive // TODO(CAPP) Fix revive linter
-func (g *dummyGauge) WithTags(tags map[string]string) telemetryComponent.SimpleGauge { return nil }
+func (g *dummyGauge) Set(float64, ...string)                                    {}
```
Comment (AliDatadog) [inline pkg/util/kubernetes/apiserver/leaderelection/leaderelection_test.go:404]: I don't know why but revive was still angry

File: pkg/clusteragent/api/leader_handler.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	// As leader, handle the request
+	lph.leaderHandler(w, r)
+}
+
+// rejectOrForwardLeaderQuery performs some checks on incoming queries that should go to a leader:
+// - Forward to leader if we are a follower
+// - Reject with "not ready" if leader election status is unknown
+func (lph *LeaderProxyHandler) rejectOrForwardLeaderQuery(rw http.ResponseWriter, req *http.Request) bool {
+	if !lph.leaderElectionEnabled {
+		return false
```
Comment (adel121) [inline pkg/clusteragent/api/leader_handler.go:93]: QQ

What happens in case leader election is disabled, and language detection is enabled? 

From what I understand from the code, I assume that language detection will not work, and we will have no indicator that the reason is that leader election is disabled. 

Would it make sense to emit an ERROR here saying that leader_election should be enabled in order for language detection requests to be handled?

File: pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go [hunk 7]
```diff
@@ -258,6 +269,8 @@ func (le *LeaderEngine) GetLeaderIP() (string, error) {
 	if err != nil {
 		return "", err
 	}
+	cache.Cache.Set(cacheKey, target.IP, 1*time.Hour)
```
Comment (AliDatadog) [inline pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go:272]: Maybe change the cache TTL in case the pod's IP changes

File: releasenotes-dca/notes/forward-pld-queries-to-leader-67ab5a941c65a47c.yaml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# Each section from every release note are combined when the
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+enhancements:
+  - |
+    Cluster-Agent Followers can now proxy patch request queries to the Leader for Library Injection with language detection.
```
Comment (davidor) [inline releasenotes-dca/notes/forward-pld-queries-to-leader-67ab5a941c65a47c.yaml:7]: I don't think this should be included in the release notes.
It's not a user-facing feature and PLD is not GA yet.

File: pkg/util/kubernetes/apiserver/leaderelection/leaderelection_test.go [hunk 9]
```diff
@@ -326,16 +324,6 @@ func TestGetLeaderIPFollower_ConfigMap(t *testing.T) {
 	ip, err := le.GetLeaderIP()
 	assert.NoError(t, err)
 	assert.Equal(t, "1.1.1.2", ip)
-
-	// Remove bar from endpoints
```
Comment (davidor) [inline pkg/util/kubernetes/apiserver/leaderelection/leaderelection_test.go:330]: Why were these removed? I think that if we clean the cache before the test it should behave as before and the test should still be valid.

File: pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go [hunk 10]
```diff
@@ -61,8 +62,11 @@ type LeaderEngine struct {
 	coordClient         coordinationv1.CoordinationV1Interface
 	ServiceName         string
 	leaderIdentityMutex sync.RWMutex
-	leaderElector       *leaderelection.LeaderElector
-	lockType            string
+	// leaderIPMutex is used to avoid concurrent calls to GetLeaderIP
+	// that would result in multiple calls to the API server
+	leaderIPMutex sync.Mutex
```
Comment (davidor) [inline pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go:67]: The cache and the mutex were not required before, why are they needed now? Is it because we expect a substantial increase in the number of request and this is needed for performance reasons?

File: pkg/clusteragent/api/leader_handler.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	if lph.le == nil {
```
Comment (davidor) [inline pkg/clusteragent/api/leader_handler.go:96]: I'm not sure why this is needed. Should we always initialize this using `WithLeaderProxyHandler` so that `leaderEngine` is always set?

Comment (AliDatadog) [reply pkg/clusteragent/api/leader_handler.go:96]: I think it's because the leader engine is not initialized when this is called. I think it's also possible that the first initialization fails and the rest work.

File: pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go (hunk 10 above)
Comment (davidor) [reply pkg/util/kubernetes/apiserver/leaderelection/leaderelection.go:67]: I think that it might not be necessary, the number of requests should be greatly reduced thanks to the cache.

File: pkg/util/kubernetes/apiserver/leaderelection/leaderelection_test.go (hunk 9 above)
Comment (AliDatadog) [reply pkg/util/kubernetes/apiserver/leaderelection/leaderelection_test.go:330]: I reintroduced them in the last commit

**Review by davidor — APPROVED** [review]

**Review by mfpierre — APPROVED** [review]
